crates/matrix-sdk-ui/tests/integration/timeline/pagination.rs

```diff
@@ -347,6 +347,112 @@ async fn dedup() {
+                &ALICE,
+                RoomMessageEventContent::text_plain("live event!"),
+            ))
+            .set_timeline_prev_batch("pagination_1".to_owned()),
+    );
+    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
+    client.sync_once(sync_settings.clone()).await.unwrap();
+    server.reset().await;
+
+    // Next sync response will response with pagination_2 token and limited
+    // response, resetting the timeline
+    sync_builder.add_joined_room(
+        JoinedRoomBuilder::new(room_id)
+            .add_timeline_event(event_builder.make_sync_message_event(
+                &BOB,
+                RoomMessageEventContent::text_plain("new live event."),
+            ))
+            .set_timeline_limited()
+            .set_timeline_prev_batch("pagination_2".to_owned()),
+    );
+    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
+
+    // pagination with first token
+    Mock::given(method("GET"))
+        .and(path_regex(r"^/_matrix/client/r0/rooms/.*/messages$"))
+        .and(header("authorization", "Bearer 1234"))
+        .and(query_param("from", "pagination_1"))
+        .respond_with(
+            ResponseTemplate::new(200)
+                .set_body_json(json!({
+                    "chunk": [],
+                    "start": "pagination_1",
+                    "end": "some_other_token",
+                }))
+                // Make sure the concurrent sync request returns first
+                .set_delay(Duration::from_millis(200)),
+        )
+        .expect(1)
+        .named("pagination_1")
+        .mount(&server)
+        .await;
+
+    // pagination with second token
+    Mock::given(method("GET"))
+        .and(path_regex(r"^/_matrix/client/r0/rooms/.*/messages$"))
+        .and(header("authorization", "Bearer 1234"))
+        .and(query_param("from", "pagination_2"))
+        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
+            "chunk": [],
+            "start": "pagination_2",
+        })))
+        .expect(1)
+        .named("pagination_2")
+        .mount(&server)
+        .await;
+
+    let mut back_pagination_status = timeline.back_pagination_status();
+
+    let paginate = async {
+        timeline
+            .paginate_backwards(PaginationOptions::single_request(10).wait_for_token())
+            .await
+            .unwrap();
+    };
+    let observe_paginating = async {
+        assert_eq!(back_pagination_status.next().await, Some(BackPaginationStatus::Paginating));
+        // timeline start reached because second pagination response contains
+        // no end field
+        assert_eq!(
+            back_pagination_status.next().await,
+            Some(BackPaginationStatus::TimelineStartReached)
+        );
+    };
+    let sync = async {
+        client.sync_once(sync_settings.clone()).await.unwrap();
+    };
+    timeout(Duration::from_secs(2), join3(paginate, observe_paginating, sync)).await.unwrap();
```

Comment:
If the timeline reset didn't happen, the amount of timeline items would be different at the end.

Reply:
Also, there wouldn't be two pagination requests, which is also tested for.

Reply:
It would also increase our test coverage :innocent:

Reply:
The documentation for `subscribe_sync_gap` says that we want to remove it in the future. I'm pretty sure we can remove it now since nothing else uses it and room update handlers fill the role of the "making event handlers more general" that that doc comment mentions.